public/app/features/explore/QueryLibrary/QueryTemplatesList.tsx

```diff
@@ -1,14 +1,101 @@
-import { EmptyState, Spinner } from '@grafana/ui';
+import { css } from '@emotion/css';
+import { uniqBy } from 'lodash';
+import { useEffect, useMemo, useState } from 'react';
+
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+import { getDataSourceSrv } from '@grafana/runtime';
+import { EmptyState, FilterInput, InlineLabel, MultiSelect, Spinner, useStyles2, Stack } from '@grafana/ui';
+import { t, Trans } from 'app/core/internationalization';
+import { createQueryText } from 'app/core/utils/richHistory';
 import { useAllQueryTemplatesQuery } from 'app/features/query-library';
 import { QueryTemplate } from 'app/features/query-library/types';
 
 import { getDatasourceSrv } from '../../plugins/datasource_srv';
 
+import { QueryLibraryProps } from './QueryLibrary';
 import QueryTemplatesTable from './QueryTemplatesTable';
 import { QueryTemplateRow } from './QueryTemplatesTable/types';
+import { searchQueryLibrary } from './utils/search';
+
+interface QueryTemplatesListProps extends QueryLibraryProps {}
 
-export function QueryTemplatesList() {
+export function QueryTemplatesList(props: QueryTemplatesListProps) {
   const { data, isLoading, error } = useAllQueryTemplatesQuery();
+  const [searchQuery, setSearchQuery] = useState('');
+  const [datasourceFilters, setDatasourceFilters] = useState<Array<SelectableValue<string>>>(
+    props.activeDatasources?.map((ds) => ({ value: ds, label: ds })) || []
+  );
+
+  const [allQueryTemplateRows, setAllQueryTemplateRows] = useState<QueryTemplateRow[]>([]);
+  const [isRowsLoading, setIsRowsLoading] = useState(true);
+  const styles = useStyles2(getStyles);
+
+  useEffect(() => {
+    let isMounted = true;
```

Comment:
The `isMounted` name is a bit misleading here as the effect runs more times than just on mount. Something like `shouldCancel` or something like that would imho more closely describe this.